sklearn/preprocessing/imputation.py

```diff
@@ -366,13 +396,66 @@ def transform(self, X):
 
             mask = _get_mask(X, self.missing_values)
             n_missing = np.sum(mask, axis=self.axis)
-            values = np.repeat(valid_statistics, n_missing)
 
-            if self.axis == 0:
-                coordinates = np.where(mask.transpose())[::-1]
+            if self.strategy == 'knn':
+                if self.axis == 1:
+                    X = X.transpose()
+                    mask = mask.transpose()
+                    statistics = statistics.transpose()
+
+                batch_size = 10  # set batch size for block query
+                if False:
+                    missing_index = np.where(mask.any(1))[0]
+                    #@jnothman 's method
+                    for sl in list(gen_batches(len(missing_index), batch_size)):
+                        index_start, index_stop = missing_index[sl][0], missing_index[sl][-1]+1
+                        X_sl = X[index_start: index_stop]
+                        mask_sl = _get_mask(X_sl, self.missing_values)
```

Comment:
Because in this version, X_sl is just a slice of original X with fixed batch_size samples, therefore it could contain samples with or without missing features. That's my original thought, but it's not a smart one. I have changed it and let X_sl only contain missing samples.